### PR TITLE
Electron layer bundling experiments

### PR DESCRIPTION
**Pull request by Nodonisko**
### Overview

This PR aims to address the recurring issue of unnecessary dependencies being included in the Electron app due to the deep nested dependency tree within our monorepo. The proposed solution is to adopt a similar approach for the Electron layer as we currently use for the UI layer of the app.

### Changes

The primary change is the creation of a new package, `@trezor/suite-desktop-core`, which is essentially a copy of the `src/` folder from `@trezor/suite-desktop` along with some necessary build scripts. The build output of this package is set to `dist/` in the existing `@trezor/suite-desktop`, and almost all dependencies from the `package.json` of the old `@trezor/suite-desktop` have been removed. The few remaining dependencies are native Electron ones that are not bundled, and these are defined as external dependencies in the build script `@trezor/suite-desktop-core/scripts/build` to prevent them from being bundled.

The `@trezor/suite-desktop` package now serves as a container for the generated/bundled code from the UI and Electron layers, and is responsible for creating the Electron app. No custom code should be added to this package, and no dependencies from the monorepo should be added to the `package.json` in `@trezor/suite-desktop`. Doing so would break this system.

### Benefits

1. App size reduction of ~6MB (measured against the latest production build .dmg file)
2. Elimination of the need to maintain an extensive files blacklist in `electron-builder-config.js`
3. Ability to use any package from the monorepo in `@trezor/suite-desktop-core` without needing to add `build:libs`
4. No need to worry about including unwanted dependencies in Electron's node_modules as the bundler will only include the code that is actually used
5. Additional layer to prevent protestware issues that we've encountered recently (styled-components, es5-ext)
6. Potential to eliminate most of the `build:libs` that were added solely for use in the Electron layer

### Drawbacks

1. Packages with binaries/node-gyp-build outputs that need to be included in the Electron app must be specifically defined. This was necessary for the `usb` package used in transport to start the app, and there may be other binaries that need to be added. If we need to add our own binaries, we should create a new package solely for binaries without any code that can be added to `@trezor/suite-desktop/package.json`.

### Next Steps

If we decide to proceed with this approach, further polishing of this PR and extensive testing will be required. I have not had the time to conduct in-depth testing, but I was able to reach the welcome screen without encountering any errors.

### Diagram

To better illustrate the changes, I will create a diagram using the Diagrams plugin. This will help to visualize the new structure and the relationship between the packages.

Diagram

P.S: This PR description was generated with big help of GPT4, let me know if you like it.

### CODE REVIEW

File: packages/suite-desktop/package.json [hunk 1]
```diff
@@ -23,47 +24,21 @@
...
         "publish:mac": "yarn build:mac && yarn build:app:electron --publish always --mac",
         "publish:win": "yarn build:win && yarn build:app:electron --publish always --win",
-        "lint:js": "yarn g:eslint '**/*.{ts,tsx,js}'",
-        "type-check": "tsc --build tsconfig.json && tsc --build scripts/tsconfig.json",
-        "test:unit": "jest",
-        "test:e2e": "yarn xvfb-maybe -- playwright test --config=./e2e/playwright.config.ts"
+        "lint:js": "yarn g:eslint '**/*.{ts,tsx,js}'"
     },
     "dependencies": {
         "@sentry/electron": "^4.6.0",
```
Comment (Nodonisko) [inline packages/suite-desktop/package.json]: could be probably removed too

**Comment by matejkriz** [comment]
> P.S: This PR description was generated with big help of GPT4, let me know if you like it.

I like the description a lot! How exactly have you used GPT4 for it? Just for polishing language and wording or did you let it read something from the code changes?

**Comment by marekrjpolak** [comment]
I report no problems with this branch on NixOS. cc: @matejkriz

**Comment by Nodonisko** [comment]
@matejkriz I wrote some very brief description myself that looked horrible and let GPT4 rewrite it and add chart.

File: packages/suite-desktop-core/package.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        "systeminformation": "^5.18.3"
+    },
+    "devDependencies": {
+        "@electron/notarize": "1.2.4",
+        "@playwright/test": "^1.35.1",
+        "@trezor/trezor-user-env-link": "workspace:*",
+        "@types/electron-localshortcut": "^3.1.0",
+        "cross-fetch": "^3.1.6",
+        "electron": "24.5.1",
+        "electron-builder": "24.4.0",
```
Comment (tomasklim) [inline packages/suite-desktop-core/package.json:51]: windows desktop app will not work with this version right? @matejkriz

File: packages/suite-desktop-core/package.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
+{
+    "name": "@trezor/suite-desktop-core",
+    "description": "Trezor Suite desktop application",
```
Comment (tomasklim) [inline packages/suite-desktop-core/package.json:3]: nit: Unnecessary description

File: packages/suite-desktop-core/README.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+### macOS
+
+`/Users/<user>/Library/Application Support/@trezor/suite-desktop/IndexedDB`
+
+### Windows
+
+`C:\Users\<user>\AppData\Roaming\@trezor\suite-desktop\IndexedDB`
+
+## Clearing Electron cache
```
Comment (tomasklim) [inline packages/suite-desktop-core/README.md]: old nit:
Should include some info about `suite-desktop`, `suite-desktop-dev` and `suite-desktop-local`

File: packages/suite-desktop-core/package.json (hunk 2 above)
Comment (tomasklim) [inline packages/suite-desktop-core/package.json]: windows desktop app will not work with this version right? @matejkriz so cannot be tested by qa

**Review by tomasklim — CHANGES_REQUESTED** [review]
just small things

File: packages/suite-desktop-core/package.json (hunk 2 above)
Comment (matejkriz) [reply packages/suite-desktop-core/package.json]: No, electron-builder is not used in suite-desktop-core and should not be here as devDependency at all... In suite-desktop, there is correct version.

**Comment by matejkriz** [comment]
QA task: https://www.notion.so/satoshilabs/Suite-Desktop-bundling-experiment-0c9a5632603443c98d6ff0ef6685d369?pvs=4

File: packages/suite-desktop/package.json (hunk 1 above)
Comment (matejkriz) [reply packages/suite-desktop/package.json]: yes, removed

File: packages/suite-desktop-core/README.md (hunk 4 above)
Comment (matejkriz) [reply packages/suite-desktop-core/README.md]: 460a69028227df7222f2b2ab4c99d5cad00b2945

**Comment by matejkriz** [comment]
We had an issue with Tor toggling that has been solved by bundling also all native modules together 0ee21f47e285b3327c04d45b97afe86bf72c7bdd. There was no reason to keep them loaded dynamically and it causes some troubles probably because of multiple instances of those modules.

File: packages/suite-desktop-core/package.json (hunk 2 above)
Comment (matejkriz) [reply packages/suite-desktop-core/package.json]: ff11317e0f71e65f10f3622e5a2630a40028e188

**Review by tomasklim — CHANGES_REQUESTED** [review]
Screenshot 2023-07-26 at 15 06 38
Is it possible to include only required platform of `usb` in `app.asar.unpacked` in `node_modules`?

**Comment by matejkriz** [comment]
> Is it possible to include only required platform of `usb` in `app.asar.unpacked` in `node_modules`?

I believe it is, but I spent more than hour trying that and found no nice solution. Worst case scenario, we can write some script that will delete other platforms, but the naming is not consistent across packages, check blake-hash...

Anyway, it's not a new issue, let's solve it separately.

**Review by tomasklim — APPROVED** [review]
LGTM

**Comment by matejkriz** [comment]
@trezor/qa please test it again in develop, because there were some following changes in native modules, so would be nice to double check 
- that Tor is working properly now - also with little snitch, 
- that everything go through tor if it's enabled, 
- how it behaves when started with tor enabled and then disabled 
- how it behaves when started with tor disabled and then enabled 
🙏

**Comment by bosomt** [comment]
**QA OK**

https://www.notion.so/satoshilabs/Suite-Desktop-bundling-experiment-volume-2-6b150084141542f3987d5dcd28dd02ec?pvs=4

**Pull request merged**